--- a/vllm/distributed/device_communicators/all2all.py
+++ b/vllm/distributed/device_communicators/all2all.py
@@ -70,6 +70,7 @@
         if extra_tensors is not None:
             tensors_to_gather.extend(extra_tensors)
 
+        dist.barrier(group=dist_group.device_group)
         gathered_tensors = dist_group.all_gatherv(
             tensors_to_gather,
             dim=0,
@@ -105,6 +106,7 @@
         if extra_tensors is not None:
             tensors_to_gather.extend(extra_tensors)
 
+        dist.barrier(group=dist_group.device_group)
         gathered_tensors = dist_group.all_gatherv(
             tensors_to_gather,
             dim=0,
@@ -132,6 +134,7 @@
         assert sizes is not None
 
         dist_group = get_ep_group() if is_sequence_parallel else get_dp_group()
+        dist.barrier(group=dist_group.device_group)
         hidden_states = dist_group.reduce_scatterv(hidden_states, dim=0, sizes=sizes)
         return hidden_states
 
